mozetl/topline/topline_summary.py

```diff
@@ -0,0 +1,295 @@
+        "client_id",
+        "timestamp",
+        "is_default_browser",
+        "search_counts",
+        "country",
+        "profile_creation_date",
+        "channel",
+        "os",
+        "hours",
+    ]
+    columns = {col: F.col(col) for col in input_columns}
+
+    # normalize countries against a whitelist
+    columns["country"] = (
+        F.when(F.col("country").isin(countries), F.col("country"))
+        .otherwise("Other")
+        .alias("country")
+    )
+
+    # clean operating system based on CEP naming scheme
+    pattern = {
+        "Windows": ["Windows%", "WINNT%"],
+        "Mac": ["Darwin%"],
+        "Linux": ["%Linux%", "%BSD%", "%SunOS%"],
+    }
+    columns["os"] = column_like("os", pattern, "Other")
+
+    # rename normalized channel to channel
+    columns["channel"] = F.col("normalized_channel")
+
+    # convert profile creation date into seconds (day -> seconds)
+    columns["profile_creation_date"] = (
+        F.when(F.col("profile_creation_date") >= 0,
+               F.col("profile_creation_date") * seconds_per_day)
+        .otherwise(0.0)
+        .cast(types.DoubleType())
+    )
+
+    # generate hours of usage from subsession length (seconds -> hours)
+    columns["hours"] = (
+        F.when((F.col("subsession_length") >= 0) &
+               (F.col("subsession_length") < 180 * seconds_per_day),
+               F.col("subsession_length") / seconds_per_hour)
+        .otherwise(0.0)
+        .cast(types.DoubleType())
+    )
+
+    # clean the dataset
+    clean = (
+        dataframe
+        .where(F.col("submission_date_s3") >= start)
+        .where(F.col("submission_date_s3") < end)
+        .select([expr.alias(name) for name, expr in columns.iteritems()])
+    )
+
+    return clean
+
+
+def search_aggregates(dataframe, attributes):
+    # search engines to pivot against
+    search_labels = ["google", "bing", "yahoo", "other"]
+
+    # patterns to filter search engines
+    patterns = {
+        "google": ["%Google%", "%google%"],
+        "bing": ["%Bing%", "%bing%"],
+        "yahoo": ["%Yahoo%", "%yahoo%"],
+    }
+
+    s_engine = (
+        column_like("search_count.engine", patterns, "other").alias("engine")
+    )
+    s_count = (
+        F.when(F.col("search_count.count") > 0, F.col("search_count.count"))
+        .otherwise(0)
+        .alias("count")
+    )
+
+    # generate the search aggregates by exploding and pivoting
+    search = (
+        dataframe
+        .withColumn("search_count", F.explode("search_counts"))
+        .select("country", "channel", "os", s_engine, s_count)
+        .groupBy(attributes)
+        .pivot("engine", search_labels)
+        .agg(F.sum("count"))
+        .na.fill(0, search_labels)
+    )
+
+    return search
+
+
+def hours_aggregates(dataframe, attributes):
+    # simple aggregate
+    hours = dataframe.groupBy(attributes).agg(F.sum("hours").alias("hours"))
+    return hours
+
+
+def client_aggregates(dataframe, timestamp, attributes):
```

Comment:
Please add a general comment about what this function is doing (get the newest record per client, determine whether the record is "new" and "default", then compute some aggregates).